README.md
# Wikimedia Maps Dashboard

This project is part of the [Discovery Dashboards](https://discovery.wmflabs.org/) project, using datasets publicly available at [analytics.wikimedia.org/datasets/discovery](https://analytics.wikimedia.org/datasets/discovery/). For more information on the datasets, refer to [README on the GitHub mirror](https://github.com/wikimedia/wikimedia-discovery-golden/blob/master/docs/README.md).

## Quick start

Install the dependencies:

```R
install.packages(c("devtools", "shiny", "reshape2", "data.table"))
devtools::install_git("https://gerrit.wikimedia.org/r/wikimedia/discovery/polloi")
```

Run the app:

```R
shiny::runApp(launch.browser = 0)
```

Please note that this project is licensed under [MIT License](LICENSE.md) and released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms. [Wikimedia technical spaces code of conduct](https://www.mediawiki.org/wiki/Special:MyLanguage/Code_of_Conduct) also applies.


CHANGELOG.md
# Change Log (Patch Notes)
All notable changes to this project will be documented in this file.

## 2017/09/18
- Modularized the dashboard source code
- Added maplink & mapframe prevalence graphs ([T170022](https://phabricator.wikimedia.org/T170022))

## 2017/06/20
- Added licensing info ([T167930](https://phabricator.wikimedia.org/T167930))

## 2017/02/02
- Updated to work with new datasets generated by Reportupdater-based golden ([T150915](https://phabricator.wikimedia.org/T150915))

## 2015/11/10
- Fixed smoothing
- Refactored large sections of code
- Added time frame selection

## 2015/09/30
- Added a change log
- Added a contributor code of conduct
- Added a readme


ui.R
library(shiny)
library(shinydashboard)
library(dygraphs)

function(request) {
  dashboardPage(

    dashboardHeader(title = "Wikimedia Maps", dropdownMenuOutput("message_menu"), disable = FALSE),

    dashboardSidebar(
      tags$head(
        tags$link(rel = "stylesheet", type = "text/css", href = "stylesheet.css"),
        tags$script(src = "custom.js")
      ),
      sidebarMenu(id = "tabs",
                  menuItem(text = "Kartotherian usage",
                           menuSubItem(text = "Summary", tabName = "tiles_summary"),
                           menuSubItem(text = "Tiles by style", tabName = "tiles_total_by_style"),
                           menuSubItem(text = "Users by style", tabName = "tiles_users_by_style"),
                           menuSubItem(text = "Tiles by zoom level", tabName = "tiles_total_by_zoom")),
                  menuItem(text = "Kartographer usage",
                           menuSubItem(text = "Overall prevalence", tabName = "kartographer_prevalence"),
                           menuSubItem(text = "Language/project breakdown", tabName = "kartographer_langproj")),
                  menuItem(text = "Platform usage", tabName = "platform_usage"),
                  menuItem(text = "Feature usage",
                           menuSubItem(text = "GeoHack", tabName = "geohack_usage"),
                           menuSubItem(text = "WikiMiniAtlas", tabName = "wikiminiatlas_usage"),
                           menuSubItem(text = "Wikivoyage", tabName = "wikivoyage_usage"),
                           menuSubItem(text = "WIWOSM", tabName = "wiwosm_usage")),
                  menuItem(text = "Geographic breakdown", tabName = "geo_breakdown"),
                  menuItem(text = "Global Settings",
                           selectInput(inputId = "smoothing_global", label = "Smoothing", selectize = TRUE, selected = "day",
                                       choices = c("No Smoothing" = "day", "Weekly Median" = "week", "Monthly Median" = "month", "Splines" = "gam")),
                           icon = icon("cog", lib = "glyphicon"))
      ),
      div(icon("info-sign", lib = "glyphicon"), HTML("<strong>Tip</strong>: you can drag on the graphs with your mouse to zoom in on a particular date range."), style = "padding: 10px; color: white;"),
      div(bookmarkButton(), style = "text-align: center;")
    ),

    dashboardBody(
      tabItems(
        tabItem(tabName = "tiles_summary",
                fluidRow(
                  column(selectInput("tiles_summary_variable", "Total", c("Users", "Tiles"), selected = "Tiles"), width = 4),
                  column(polloi::smooth_select("smoothing_tiles_summary_series"), width = 4),
                  column(
                    HTML("<label class = \"control-label\" style = \"margin-bottom:-30px;\">Scaling</label>"),
                    checkboxInput("tiles_summary_logscale", "Use Log scale", FALSE),
                    width = 4
                  )
                ),
                polloi::automata_select("tile_summary_automata_check"),
                dygraphOutput("tiles_summary_series"),
                div(id = "tiles_summary_series_legend", style = "text-align: right;"),
                includeMarkdown("./tab_documentation/tiles_summary.md")),
        tabItem(tabName = "tiles_total_by_style",
                fluidRow(
                  column(polloi::smooth_select("smoothing_tiles_style_series"), width = 4),
                  column(
                    HTML("<label class = \"control-label\" style = \"margin-bottom:-30px;\">Scaling</label>"),
                    checkboxInput("tiles_style_logscale", "Use Log scale", FALSE),
                    width = 4
                  )
                ),
                polloi::automata_select("tile_style_automata_check"),
                dygraphOutput("tiles_style_series"),
                div(id = "tiles_style_series_legend", style = "text-align: right;"),
                includeMarkdown("./tab_documentation/tiles_total_by_style.md")),
        tabItem(tabName = "tiles_users_by_style",
                fluidRow(
                  column(polloi::smooth_select("smoothing_tiles_users_series"), width = 4),
                  column(
                    HTML("<label class = \"control-label\" style = \"margin-bottom:-30px;\">Scaling</label>"),
                    checkboxInput("tiles_users_logscale", "Use Log scale", FALSE),
                    width = 4
                  )
                ),
                polloi::automata_select("tile_users_automata_check"),
                dygraphOutput("tiles_users_series"),
                div(id = "tiles_users_series_legend", style = "text-align: right;"),
                includeMarkdown("./tab_documentation/tiles_users_by_style.md")),
        tabItem(tabName = "tiles_total_by_zoom",
                fluidRow(
                  column(polloi::smooth_select("smoothing_tiles_zoom_series"), width = 4),
                  column(
                    HTML("<label class = \"control-label\" style = \"margin-bottom:-30px;\">Scaling</label>"),
                    checkboxInput("tiles_zoom_logscale", "Use Log scale", FALSE),
                    width = 4
                  )
                ),
                fluidRow(
                  column(uiOutput("zoom_level_selector_container"), width = 3),
                  column(dygraphOutput("tiles_zoom_series"), width = 8)
                ),
                polloi::automata_select("tile_zoom_automata_check"),
                div(id = "tiles_zoom_series_legend", style = "text-align: right;"),
                includeMarkdown("./tab_documentation/tiles_total_by_zoom.md")),
        tabItem(tabName = "kartographer_prevalence",
                dygraphOutput("overall_prevalence_series"),
                div(id = "overall_prevalence_series_legend", style = "text-align: right;"),
                includeMarkdown("./tab_documentation/overall_prevalence.md")),
        tabItem(tabName = "kartographer_langproj",
                fluidRow(
                  column(polloi::smooth_select("smoothing_prevalence_langproj"), width = 4),
                  column(checkboxGroupInput(
                    "prevalence_langproj_feature", "Feature",
                    c("Mapframe", "Maplink"), inline = TRUE,
                    selected = c("Mapframe", "Maplink")
                  ), width = 4),
                  column(radioButtons(
                    "prevalence_langproj_aggregation", "Aggregation",
                    c("Overall", "Average", "Median"), inline = TRUE,
                    selected = "Average"
                  ), width = 4)
                ),
                fluidRow(
                  column(
                    selectInput("project_order", "Sort projects by",
                                list("Alphabetical order" = "alphabet", "Volume of articles" = "volume"),
                                selected = "volume"),
                    uiOutput("project_selector_container"),
                    width = 2
                  ),
                  column(
                    selectInput("language_order", "Sort languages by",
                                list("Alphabetical order" = "alphabet", "Volume of articles" = "volume"),
                                selected = "volume"),
                    uiOutput("language_selector_container"),
                    width = 2
                  ),
                  column(
                    dygraphOutput("prevalence_langproj_plot"),
                    div(id = "prevalence_langproj_legend", style = "text-align: right;"),
                    width = 8
                  )
                ),
                includeMarkdown("./tab_documentation/prevalence_langproj.md")),
        tabItem(tabName = "platform_usage",
                fluidRow(
                  column(polloi::smooth_select("smoothing_users_per_platform"), width = 3),
                  column(checkboxInput("users_per_platform_logscale", "Use Log scale", FALSE), width = 3)
                ),
                dygraphOutput("users_per_platform"),
                div(id = "users_per_platform_legend", style = "text-align: right;"),
                includeMarkdown("./tab_documentation/unique_users.md")),
        tabItem(tabName = "geohack_usage",
                fluidRow(
                  column(polloi::smooth_select("smoothing_geohack_feature_usage"), width = 4),
                  column(checkboxInput("geohack_feature_usage_logscale", "Use Log scale", FALSE), width = 4)
                ),
                dygraphOutput("geohack_feature_usage"),
                includeMarkdown("./tab_documentation/geohack_usage.md")),
        tabItem(tabName = "wikiminiatlas_usage",
                fluidRow(
                  column(polloi::smooth_select("smoothing_wikiminiatlas_feature_usage"), width = 4),
                  column(checkboxInput("wikiminiatlas_feature_usage_logscale", "Use Log scale", FALSE), width = 4)
                ),
                dygraphOutput("wikiminiatlas_feature_usage"),
                includeMarkdown("./tab_documentation/wikiminiatlas_usage.md")),
        tabItem(tabName = "wikivoyage_usage",
                fluidRow(
                  column(polloi::smooth_select("smoothing_wikivoyage_feature_usage"), width = 4),
                  column(checkboxInput("wikivoyage_feature_usage_logscale", "Use Log scale", FALSE), width = 4)
                ),
                dygraphOutput("wikivoyage_feature_usage"),
                includeMarkdown("./tab_documentation/wikivoyage_usage.md")),
        tabItem(tabName = "wiwosm_usage",
                fluidRow(
                  column(polloi::smooth_select("smoothing_wiwosm_feature_usage"), width = 4),
                  column(checkboxInput("wiwosm_feature_usage_logscale", "Use Log scale", FALSE), width = 4)
                ),
                dygraphOutput("wiwosm_feature_usage"),
                includeMarkdown("./tab_documentation/wiwosm_usage.md")),
        tabItem(tabName = "geo_breakdown",
                polloi::smooth_select("smoothing_users_by_country"),
                dygraphOutput("users_by_country"),
                includeMarkdown("./tab_documentation/geo_breakdown.md"))
      )
    ),

    skin = "green", title = "Wikimedia Maps Usage Dashboard | Audiences | Wikimedia Foundation")
}


utils.R
# Dependencies
library(magrittr)

# Read in action data and write it into the global scope with sufficient formatting to be trivially
# used in dygraphs.
read_actions <- function() {
  # Read in data, split and rename it, and write it o
  usage_data <<- polloi::read_dataset("discovery/metrics/maps/actions_per_tool.tsv", col_types = "Dcci") %>%
    dplyr::filter(!is.na(feature), !is.na(action), !is.na(events)) %>%
    { split(.[, c(1, 3, 4)], .$feature) } %>%
    lapply(tidyr::spread_, key = "action", value = "events", fill = 0)
  return(invisible())
}

# Read in user count
read_users <- function() {
  user_data <<- polloi::read_dataset("discovery/metrics/maps/users_per_feature.tsv", col_types = "Dci") %>%
    dplyr::filter(!is.na(feature), !is.na(users)) %>%
    tidyr::spread(feature, users, fill = 0)
  return(invisible())
}

read_tiles <- function() {
  new_tiles_automata <<- polloi::read_dataset("discovery/metrics/maps/tile_aggregates_with_automata.tsv", col_types = "Dcidcciidiii") %>%
    dplyr::filter(!is.na(total), !is.na(users)) %>%
    data.table::as.data.table()
  new_tiles_no_automata <<- polloi::read_dataset("discovery/metrics/maps/tile_aggregates_no_automata.tsv", col_types = "Dcidcciidiii") %>%
    dplyr::filter(!is.na(total), !is.na(users)) %>%
    data.table::as.data.table()
  return(invisible())
}

read_countries <- function(){
  country_data <<- polloi::read_dataset("discovery/metrics/maps/users_by_country.tsv", col_types = "Dcd") %>%
    dplyr::filter(!is.na(country), !is.na(users)) %>%
    dplyr::mutate(users = users * 100) %>%
    tidyr::spread(country, users, fill = 0)
  return(invisible())
}


modules/feature_usage.R
output$users_per_platform <- renderDygraph({
  user_data %>%
    polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_users_per_platform)) %>%
    polloi::make_dygraph("Date", "Events", "Unique users by platform, by day") %>%
    dyAxis("y", logscale = input$users_per_platform_logscale) %>%
    dyLegend(labelsDiv = "users_per_platform_legend", show = "always") %>%
    dyRangeSelector %>%
    dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
})

output$geohack_feature_usage <- renderDygraph({
  usage_data$GeoHack %>%
    polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_geohack_feature_usage)) %>%
    polloi::make_dygraph("Date", "Events", "Feature usage for GeoHack") %>%
    dyRangeSelector %>%
    dyAxis("y", logscale = input$geohack_feature_usage_logscale) %>%
    dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
})

output$wikiminiatlas_feature_usage <- renderDygraph({
  usage_data$WikiMiniAtlas %>%
    polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_wikiminiatlas_feature_usage)) %>%
    polloi::make_dygraph("Date", "Events", "Feature usage for WikiMiniAtlas") %>%
    dyRangeSelector %>%
    dyAxis("y", logscale = input$wikiminiatlas_feature_usage_logscale) %>%
    dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
})

output$wikivoyage_feature_usage <- renderDygraph({
  usage_data$Wikivoyage %>%
    polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_wikivoyage_feature_usage)) %>%
    polloi::make_dygraph("Date", "Events", "Feature usage for Wikivoyage") %>%
    dyRangeSelector %>%
    dyAxis("y", logscale = input$wikivoyage_feature_usage_logscale) %>%
    dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
})

output$wiwosm_feature_usage <- renderDygraph({
  usage_data$WIWOSM %>%
    polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_wiwosm_feature_usage)) %>%
    polloi::make_dygraph("Date", "Events", "Feature usage for WIWOSM") %>%
    dyRangeSelector %>%
    dyAxis("y", logscale = input$wiwosm_feature_usage_logscale) %>%
    dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
    dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
})


server.R
source("utils.R")

existing_date <- Sys.Date() - 1

# Actual server code.
shinyServer(function(input, output, session) {

  if (Sys.Date() != existing_date) {
    # Create a Progress object
    progress <- shiny::Progress$new()
    progress$set(message = "Downloading EL actions data", value = 0)
    read_actions()
    progress$set(message = "Downloading EL user counts", value = 0.2)
    read_users()
    progress$set(message = "Downloading tile usage data", value = 0.4)
    suppressWarnings(read_tiles())
    progress$set(message = "Downloading geography data", value = 0.8)
    read_countries()
    progress$set(message = "Finished downloading datasets", value = 1)
    existing_date <<- Sys.Date()
    progress$close()
  }

  output$tiles_summary_series <- renderDygraph({
    temp <- polloi::data_select(input$tile_summary_automata_check, new_tiles_automata, new_tiles_no_automata)[, list(
      `total tiles` = sum(total),
      `total users` = sum(users),
      `average tiles per user` = sum(total)/sum(users)
    ), by = "date"]
    switch(input$tiles_summary_variable,
           Users = { temp %<>% dplyr::select(-`total tiles`) },
           Tiles = { temp %<>% dplyr::select(-`total users`) })
    temp %<>% polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_tiles_summary_series))
    polloi::make_dygraph(temp, "Date", input$tiles_summary_variable, "Tile usage") %>%
      dySeries(name = grep('average tiles per user', names(temp), value = TRUE), axis = 'y2') %>%
      dyAxis(name = 'y', drawGrid = FALSE, logscale = input$tiles_summary_logscale) %>%
      dyAxis(name = 'y2', independentTicks = TRUE, drawGrid = FALSE) %>%
      dyLegend(labelsDiv = "tiles_summary_series_legend", show = "always") %>%
      dyRangeSelector(retainDateWindow = TRUE) %>%
      dyEvent(as.Date("2015-09-17"), "A (announcement)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-01-08"), "B (enwiki launch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-01-12"), "C (cache clear)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-11-09"), "D (pkget)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-08-14"), "E (pkget)", labelLoc = "bottom")
  })

  output$tiles_style_series <- renderDygraph({
    polloi::data_select(
      input$tile_style_automata_check,
      new_tiles_automata,
      new_tiles_no_automata
    )[, j = list(`total tiles` = sum(total)),
        by = c("date", "style")] %>%
      tidyr::spread(style, `total tiles`, fill = 0) %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_tiles_style_series)) %>%
      polloi::make_dygraph("Date", "Tiles", "Total tiles by style", legend_name = "Style") %>%
      dyAxis("y", logscale = input$tiles_style_logscale) %>%
      dyLegend(labelsDiv = "tiles_style_series_legend", show = "always") %>%
      dyRangeSelector %>%
      dyEvent(as.Date("2015-09-17"), "A (announcement)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-01-08"), "B (enwiki launch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-01-12"), "C (cache clear)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-11-09"), "D (pkget)", labelLoc = "top") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-08-14"), "E (pkget)", labelLoc = "bottom")
  })

  output$tiles_users_series <- renderDygraph({
    polloi::data_select(
      input$tile_users_automata_check,
      new_tiles_automata,
      new_tiles_no_automata
    )[, j = list(`total users` = sum(users)),
        by = c("date", "style")] %>%
      tidyr::spread(style, `total users`, fill = 0) %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_tiles_users_series)) %>%
      polloi::make_dygraph("Date", "Users", "Total users by style") %>%
      dyAxis("y", logscale = input$tiles_users_logscale) %>%
      dyLegend(labelsDiv = "tiles_users_series_legend", show = "always") %>%
      dyRangeSelector %>%
      dyEvent(as.Date("2015-09-17"), "A (announcement)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-01-08"), "B (enwiki launch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-01-12"), "C (cache clear)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-11-08"), "D (pkget)", labelLoc = "top") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-08-14"), "E (pkget)", labelLoc = "bottom")
  })

  output$zoom_level_selector_container <- renderUI({
    selectInput("zoom_level_selector", "Zoom level",
                multiple = TRUE, selected = "0", selectize = FALSE, size = 19,
                choices = as.character(sort(unique(new_tiles_no_automata$zoom))))
  })

  output$tiles_zoom_series <- renderDygraph({
    req(input$zoom_level_selector)
    polloi::data_select(
      input$tile_zoom_automata_check,
      new_tiles_automata,
      new_tiles_no_automata
    )[zoom %in% as.numeric(input$zoom_level_selector),
      j = list(`total tiles` = sum(total)),
      by = c("date", "zoom")] %>%
      tidyr::spread(zoom, `total tiles`, fill = 0) %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_tiles_zoom_series)) %>%
      polloi::make_dygraph("Date", "Tiles", "Total tiles by zoom level") %>%
      dyAxis("y", logscale = input$tiles_zoom_logscale) %>%
      dyLegend(labelsDiv = "tiles_zoom_series_legend", show = "always") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  })

  output$users_per_platform <- renderDygraph({
    user_data %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_users_per_platform)) %>%
      polloi::make_dygraph("Date", "Events", "Unique users by platform, by day") %>%
      dyAxis("y", logscale = input$users_per_platform_logscale) %>%
      dyLegend(labelsDiv = "users_per_platform_legend", show = "always") %>%
      dyRangeSelector %>%
      dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  })

  output$geohack_feature_usage <- renderDygraph({
    usage_data$GeoHack %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_geohack_feature_usage)) %>%
      polloi::make_dygraph("Date", "Events", "Feature usage for GeoHack") %>%
      dyRangeSelector %>%
      dyAxis("y", logscale = input$geohack_feature_usage_logscale) %>%
      dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  })

  output$wikiminiatlas_feature_usage <- renderDygraph({
    usage_data$WikiMiniAtlas %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_wikiminiatlas_feature_usage)) %>%
      polloi::make_dygraph("Date", "Events", "Feature usage for WikiMiniAtlas") %>%
      dyRangeSelector %>%
      dyAxis("y", logscale = input$wikiminiatlas_feature_usage_logscale) %>%
      dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  })

  output$wikivoyage_feature_usage <- renderDygraph({
    usage_data$Wikivoyage %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_wikivoyage_feature_usage)) %>%
      polloi::make_dygraph("Date", "Events", "Feature usage for Wikivoyage") %>%
      dyRangeSelector %>%
      dyAxis("y", logscale = input$wikivoyage_feature_usage_logscale) %>%
      dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  })

  output$wiwosm_feature_usage <- renderDygraph({
    usage_data$WIWOSM %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_wiwosm_feature_usage)) %>%
      polloi::make_dygraph("Date", "Events", "Feature usage for WIWOSM") %>%
      dyRangeSelector %>%
      dyAxis("y", logscale = input$wiwosm_feature_usage_logscale) %>%
      dyEvent(as.Date("2016-04-15"), "A (Maps EL bug)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-06-17"), "A (Maps EL patch)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  })


  output$users_by_country <- renderDygraph({
    country_data %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_users_by_country)) %>%
      polloi::make_dygraph("Date", "Users (%)", "Geographic breakdown of maps users") %>%
      dyRangeSelector(fillColor = "", strokeColor = "") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  })

  # Check datasets for missing data and notify user which datasets are missing data (if any)
  output$message_menu <- renderMenu({
    notifications <- list(
      polloi::check_yesterday(usage_data[[1]], "action data"),
      polloi::check_past_week(usage_data[[1]], "action data"),
      polloi::check_yesterday(user_data, "user counts"),
      polloi::check_past_week(user_data, "user counts"),
      polloi::check_yesterday(new_tiles_automata, "tile usage data"),
      polloi::check_past_week(new_tiles_automata, "tile usage data"))
    notifications <- notifications[!sapply(notifications, is.null)]
    return(dropdownMenu(type = "notifications", .list = notifications, badgeStatus = "warning"))
  })

})
